File: 2019/ruby/lib/aoc2019/day01.rb
module Aoc2019
  class Day01
    def self.line(input)
      (input / 3).round(half: :down) - 2
    end

    def self.part_one(input)
      input.split.map { |x| line(x.to_i) }.sum
    end

    def self.line2(mass, sum = 0)
      new_mass = line(mass)
      return sum if new_mass <= 0

      line2(new_mass, sum + new_mass)
    end

    def self.part_two(input)
      input.split.map { |x| line2(x.to_i) }.sum
    end
  end
end

File: 2019/elixir/lib/Aoc2019/Day01.ex
defmodule Aoc2019.Day01 do
  defp parse(input) do
    input
    |> String.split("\n")
    |> Enum.reject(fn x -> x == "" end)
    |> Enum.map(&String.to_integer/1)
  end

  def part_one(input) do
    input
    |> parse
    |> Enum.map(&mass/1)
    |> Enum.sum()
  end

  def mass(m) do
    trunc(m / 3) - 2
  end

  def part_two(input) do
    input
    |> parse
    |> Enum.map(&culm_mass/1)
    |> Enum.sum()
  end

  def culm_mass(m), do: culm_mass(m, 0)

  def culm_mass(m, sum) do
    case mass(m) do
      x when x <= 0 -> sum
      x -> culm_mass(x, x + sum)
    end
  end
end
